libraries/Microsoft.Bot.Builder/OAuth/OAuthMessageClient.cs

```diff
@@ -0,0 +1,170 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.OAuth
+{
+    /// <summary>
+    /// Used for properly constructing a User OAuth message with OAuthCard of the channel supports it, or SignInCard.
+    /// </summary>
+    public class OAuthMessageClient
+    {
+        private readonly OAuthSettings _settings;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OAuthMessageClient"/> class.
+        /// </summary>
+        public OAuthMessageClient() 
+        { 
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="OAuthMessageClient"/> class.
+        /// </summary>
+        /// <param name="settings">Settings specific to this <see cref="OAuthMessageClient"/>.</param>
+        public OAuthMessageClient(OAuthSettings settings) 
+        {
+            _settings = settings ?? throw new NullReferenceException(nameof(settings));
+        }
+
+        /// <summary>
+        /// This method will properly construct an OAuthCard based on <see cref="OAuthSettings"/>
+        /// provided during class construction, or from the <paramref name="prompt"/> if present.
+        /// </summary>
+        /// <param name="activity">The incoming <see cref="Activity"/> to use while constructing the card and response activity.</param>
+        /// <param name="userTokenClient">The <see cref="UserTokenClient"/> to use for retrieving the <see cref="SignInResource"/> for the card.</param>
+        /// <param name="settings"><see cref="OAuthSettings"/> to use while constructing the OAuthCard.</param>
+        /// <param name="prompt"><see cref="IMessageActivity"/> to use for sending the OAuthCard. If this activity 
+        /// does not have an attachment containing an OAuthCard, then one will be added.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to use for async operations from this method.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task<IMessageActivity> GetCardMessageFromActivityAsync(Activity activity, UserTokenClient userTokenClient, OAuthSettings settings = default(OAuthSettings), IMessageActivity prompt = default(IMessageActivity), CancellationToken cancellationToken = default(CancellationToken))
```

Comment:
Do we need to return IMessageActivity? Or just Activity would do? 
Also, if we rename this class as XYZFactory, let's call this method CreateOAuthActivityAsync or something like that (is not getting the activity for use, is creating it)

Reply:
fixed